### PR TITLE
[SYCL] Fix host_image_accessor_read

### PR DESCRIPTION
**Pull request by romanovvlad**
Avoid having two host accessor simultaneously.

Signed-off-by: Vlad Romanov <vlad.romanov@intel.com>

### CODE REVIEW

File: sycl/test/basic_tests/host_image_accessor_read.cpp [hunk 1]
```diff
@@ -26,50 +26,62 @@ int foo(float *image_data) {
...
+    }
 
-    // Test image read with sampler.
-    cl::sycl::cl_float4 Ret_data2 = host_image_acc.read(Coords, Sampler);
-    assert((float)Ret_data2.x() == 85);
-    assert((float)Ret_data2.y() == 86);
-    assert((float)Ret_data2.z() == 87);
-    assert((float)Ret_data2.w() == 88);
+    {
+      auto host_image_acc2 =
```
Comment (sergey-semenov) [inline sycl/test/basic_tests/host_image_accessor_read.cpp]: Nitpick: with the addition of scopes we can probably do away with the numbering of these local variables across different scopes (host_image_acc2, Sampler2).

Comment (romanovvlad) [reply sycl/test/basic_tests/host_image_accessor_read.cpp]: Done.

File: sycl/test/basic_tests/host_image_accessor_read.cpp [hunk 2]
```diff
@@ -22,54 +22,66 @@ int foo(float *image_data) {
...
-        cl::sycl::coordinate_normalization_mode::unnormalized,
-        cl::sycl::addressing_mode::clamp_to_edge, cl::sycl::filtering_mode::nearest);
-    cl::sycl::cl_int4 OutBnds_Coords{2, 2, 3, 0};
-    cl::sycl::cl_float4 OutBnds_RetData = host_image_acc.read(OutBnds_Coords, Sampler2);
-    assert((float)OutBnds_RetData.x() == 105);
-    assert((float)OutBnds_RetData.y() == 106);
-    assert((float)OutBnds_RetData.z() == 107);
-    assert((float)OutBnds_RetData.w() == 108);
+      // Test Out-of-bounds access for clamp_to_edge Addressing Mode.
+      auto Sampler2 = cl::sycl::sampler(
```
Comment (sergey-semenov) [inline sycl/test/basic_tests/host_image_accessor_read.cpp]:
```suggestion
      auto Sampler = cl::sycl::sampler(
```

Comment (romanovvlad) [reply sycl/test/basic_tests/host_image_accessor_read.cpp]: Done.

**Comment by garimagu** [comment]
Why do you need to avoid using two host image accessors at the same time?

**Comment by romanovvlad** [comment]
> 
> 
> Why do you need to avoid using two host image accessors at the same time?
> You mean more than 1 to the same image?

Yes, I mean more than 1 to the same image.

**Comment by romanovvlad** [comment]
> 
> 
> > Why do you need to avoid using two host image accessors at the same time?
> > You mean more than 1 to the same image?
> 
> Yes, I mean more than 1 to the same image.

Is it OK with you?

**Comment by garimagu** [comment]
> > > Why do you need to avoid using two host image accessors at the same time?
> > > You mean more than 1 to the same image?
> > 
> > 
> > Yes, I mean more than 1 to the same image.
> 
> @garimagu Is it OK with you?

I do not understand why this change is needed. Was it causing any failures?
The spec doesn't stop you from doing it. Does it?
Since the spec did not give read_write access to a host image until sometime ago ; if a kernel wants to read and write to the same image, the only way to do that is using two accessors to the same image.

**Comment by romanovvlad** [comment]
> 
> 
> > > > Why do you need to avoid using two host image accessors at the same time?
> > > > You mean more than 1 to the same image?
> > > 
> > > 
> > > Yes, I mean more than 1 to the same image.
> > 
> > 
> > @garimagu Is it OK with you?
> 
> I do not understand why this change is needed. Was it causing any failures?
> The spec doesn't stop you from doing it. Does it?
> Since the spec did not give read_write access to a host image until sometime ago ; if a kernel wants to read and write to the same image, the only way to do that is using two accessors to the same image.

This is not causing failures because we have a bug in SYCL implementation.
Probably there is more clear words in the spec, but what I found:

> A special case of requirement is the one defined by a host accessor. Host accessors are represented with H(MemoryObjectaccessMode), e.g, H(b1RW) represents a host accessor to b1 in read-write mode. Host accessors are a special type of accessor constructed from a memory object outside a command group, and require that the data associated with the given memory object is available on the host in the given pointer. This causes the runtime to block on construction of this object until the requirement has been satisfied. Host accessor objects are effectively barriers on all accesses to a certain memory object.

> Since the spec did not give read_write access to a host image until sometime ago ; if a kernel wants to read and write to the same image, the only way to do that is using two accessors to the same image.

The patch changes how host accessors are used, not ones that are used in a kernel.

**Comment by garimagu** [comment]
Thanks for the explanation. Looks good to me.

**Pull request merged**